nntrainer/include/properties.h

```diff
@@ -0,0 +1,351 @@
+ *  If you want to add a property, only this part need change.               *
+ *  You only have to add property struct and prop_traits.                    *
+ *                                                                           *
+ *  It is recommened to use unique key for each types.                       *
+ *  However, it is allowed to use duplicated key                             *
+ *  When those properties are mutually exclusive                             *
+ *****************************************************************************/
+
+/**< color is here for demonstration purpose. */
+
+enum class Color {
+  black,
+  white,
+  unknown,
+};
+
+/**
+ * @brief trait specialization part
+ */
+template <> struct prop_traits<Color> {
+
+  static constexpr bool is_property = true;
+
+  typedef Color value_type;
+
+  static std::string getKey() { return "color"; }
+
+  static bool is_valid(Color c) { return c != Color::unknown; }
+
+  static Color getDefault() { return Color::unknown; }
+
+  static std::string serialize(Color c) {
+    switch (c) {
+    case Color::black:
+      return "black";
+    case Color::white:
+      return "white";
+    default:
+      return "unknown";
+    }
+  }
+
+  static Color deserialize(const std::string &val) {
+    if (val == "black") {
+      return Color::black;
+    }
+    if (val == "white") {
+      return Color::white;
+    }
+    return Color::unknown;
+  }
+};
+
+/*****************************************************************************
+ *  End of property registration part.                                       *
+ *****************************************************************************/
+
+/**
+ * @brief base class of property holder. This property holder enables set/get
+ * access to props by string
+ */
+class PropertyHolder {
+
+public:
+  /**
+   * @brief virtual destructor of class
+   */
+  virtual ~PropertyHolder() {}
+
+  /**
+   * @brief setProperty from string representation of value
+   * @param[in] value to be deserialized by prop trait.
+   * @throw std::invalid_argument if validation fails
+   */
+  virtual void setProp(const std::string &val) = 0;
+
+  /**
+   * @brief getProperty to string representation of value
+   * @retval string reprenstation of current value
+   */
+  virtual std::string getProp() = 0;
+};
+
+/**
+ * @brief actual holder of property. This class holds pointer of actual type.
+ * Which will be residing on Properties::items;
+ */
+template <typename T> class _PropertyHolder : public PropertyHolder {
+  static_assert(prop_traits<T>::is_property, "Only property type can be used");
+
+public:
+  typedef typename prop_traits<T>::value_type value_type;
+  typedef T prop_type;
+
+  /**
+   * @brief property holder destructor
+   */
+  ~_PropertyHolder() {}
+
+  _PropertyHolder() {}
+
+  value_type get() { return data; }
+
+  void set(value_type _data) {
+    if (prop_traits<T>::is_valid(_data) == false) {
```

Comment:
how about update the return value of `is_valid` to be true when valid or rename to `is_invalid`?

Reply:
`is_valid` returns true already. only undefined types return `std::runtime_error`